server/src-lib/Hasura/App.hs

```diff
@@ -632,42 +643,58 @@ instance HttpLog AppM where
     unLogger logger $ mkHttpLog $
       mkHttpAccessLogContext userInfoM reqId waiReq compressedResponse qTime cType headers
 
-instance MonadExecuteQuery AppM where
+instance MonadExecuteQuery ServerAppM where
   cacheLookup _ _ = pure ([], Nothing)
   cacheStore  _ _ = pure ()
 
-instance UserAuthentication (Tracing.TraceT AppM) where
+instance UserAuthentication (Tracing.TraceT ServerAppM) where
   resolveUserInfo logger manager headers authMode =
     runExceptT $ getUserInfoWithExpTime logger manager headers authMode
 
-instance MetadataApiAuthorization AppM where
+instance MetadataApiAuthorization ServerAppM where
   authorizeMetadataApi query userInfo = do
     let currRole = _uiRole userInfo
     when (requiresAdmin query && currRole /= adminRoleName) $
       withPathK "args" $ throw400 AccessDenied errMsg
     where
       errMsg = "restricted access : admin only"
 
-instance ConsoleRenderer AppM where
+instance ConsoleRenderer ServerAppM where
   renderConsole path authMode enableTelemetry consoleAssetsDir =
     return $ mkConsoleHTML path authMode enableTelemetry consoleAssetsDir
 
-instance MonadGQLExecutionCheck AppM where
+instance MonadGQLExecutionCheck ServerAppM where
   checkGQLExecution userInfo _ enableAL sc query = runExceptT $ do
     req <- toParsed query
     checkQueryInAllowlist enableAL userInfo req sc
     return req
 
-instance MonadConfigApiHandler AppM where
+instance MonadConfigApiHandler ServerAppM where
   runConfigApiHandler = configApiGetHandler
 
-instance MonadQueryLog AppM where
+instance MonadQueryLog ServerAppM where
   logQueryLog logger query genSqlM reqId =
     unLogger logger $ QueryLog query genSqlM reqId
 
-instance WS.MonadWSLog AppM where
+instance WS.MonadWSLog ServerAppM where
   logWSLog = unLogger
 
+runTxInMetadataStorage :: Q.TxE QErr a -> MetadataStorageT ServerAppM a
+runTxInMetadataStorage tx = do
+  pool <- lift ask
+  liftEitherM $ liftIO $ runExceptT $
+    Q.runTx pool (Q.RepeatableRead, Just Q.ReadWrite) tx
+
+instance MonadMetadataStorage (MetadataStorageT ServerAppM) where
```

Comment:
The way these pieces fit together seems a little unusual to me:

  * It’s odd to see an instance like `instance MonadFoo (BarT BazM)` rather than `instance MonadFoo BazM` or `instance MonadBaz m => MonadFoo (BarT m)`. That is, you normally define instances on a concrete monad or on a transformer with a polymorphic monad underneath, not on a transformer with a concrete monad underneath.

  * `MetadataStorageT` sounds like it’s *the* transformer for metadata storage, but it looks like this instance really implements “Postgres metadata storage.”

  * The use of the `runTxInMetadataStorage` wrapper—rather than just using `liftTx`—is a bit opaque to me. (This also seems like it will run each of these methods in a totally separate transaction—is that desirable in general?)

Reasoning through the surrounding context, it seems like perhaps a couple concerns have been muddled a bit:

  1. We need a way to acquire a Postgres connection.

  2. We need a way to delimit the scope of a Postgres transaction.

  3. We need a way to handle errors that result from executing queries against Postgres.

  4. We need a way to select which implementation of metadata storage we want to use.

`MetadataStorageT` is a bit odd as-written, because it appears to be doing points 2 through 4, which in my mind are not really usefully coupled concerns. One could imagine having things that aren’t related to metadata storage that still want to raise errors; those errors should probably go to the same place as the metadata storage errors, but it seems odd for `MetadataStorageT` to be that place. Similarly, it’s not clear to me that the scope of a Postgres transaction should be limited to a single method of `MonadMetadataStorage`.

It’s worth thinking about what exactly `MetadataStorageT`’s purpose is. If it’s just a compile-time constant of sorts that’s used to guide typeclass instance selection, maybe it should be the identity transformer:

```haskell
newtype PGMetadataStorageT m a
  = PGMetadataStorageT { runPGMetadataStorageT :: m a }
```

Then you can hang a typeclass instance on it that looks like this:

```haskell
instance MonadTx m => MonadMetadataStorage (PGMetadataStorageT m) where
  getDeprivedCronTriggerStats = getDeprivedCronTriggerStatsTx
  getScheduledEventsForDelivery = getScheduledEventsForDeliveryTx
  insertScheduledEvent = insertScheduledEventTx
  ...
```

This seems like a division of responsibilities that would make more sense to me. Concerns 2 and 3 can be handled by some other transformer in the stack, which makes a bit more sense, since they’re all very closely related.

Reply:
@lexi-lambda 
> This also seems like it will run each of these methods in a totally separate transaction?

Yes.

> Is that desirable in general?

The Metadata storage abstraction need NOT to be tied to Postgres. That said, it is the implementer choice to use any storage. In OSS, we choose that to be Postgres. 

In a nutshell the `MonadMetadataStorage` have functions in which each of them should operate in `MonadError QErr m`. Maybe the name `MetadataStorageT` is mis-leading but it just a wrapper over `ExceptT` so as to fulfill the `MonadError QErr` constraint here.

> not clear to me that the scope of a Postgres transaction should be limited to a single method of  `MonadMetadataStorage`

We're implementing the eventing sub-system for scheduled triggers in metadata storage. Each method is an isolated interaction to the storage and it is not necessary to guarantee a transaction.

Reply:
Probably it is also worth mentioning that the existing `AppM` type has been renamed to `ServerAppM`. I am not sure of the reasoning (maybe @rakeshkky can shed some light), but maybe something to keep in mind that this `ServerAppM` is not a newly introduced type but the existing app-monad type we had.

Reply:
@ecthiender The `ServerAppM` is specifically where `runHGEServer` function operates in. Also, now it carries the reader context of metadata storage pool (`Q.PGPool`).

Reply:
Would something along the following lines work and/or make sense, @rakeshkky?

```suggestion
instance (MonadReader Q.PGPool m, MonadTx m) => MonadMetadataStorage (MetadataStorageT m) where
```

Reply:
@abooij I think we shouldn't limit MetadataStorage implementation to Postgres. Hence we don't want to put `MonadReader Q.PGPool` constraint.

Reply:
@abooij Deriving `MonadMetadataStorage` instance for any generalized `MetadataStorageT m` raises overlapping instance if Pro/Cloud wants to have their own implementation of the same.

Reply:
I think the idea that we have a shared `MetadataStorageT` for which different implementations provide instances like `MonadMetadataStorage (MetadataStorageT SomeBaseMonad)` is suspect. Generally, the idea would be that each implementation of `MonadMetadataStorage` would use a *different* monad transformer. This makes sense, because the transformer might need to have a different representation—currently `MetadataStorageT` is just a wrapper around `ExceptT`, and some other `MonadMetadataStorage` instance might need to track some other information in their metadata storage transformer.

What you say about each method running in its own transaction makes sense; that is all reasonable. But I still think the other points from my comment are relevant. The idea of “bracketing a Postgres transaction,” “implementing the metadata operations,” and “having a place to raise errors to” are all independent concerns. For example, suppose in the Haskell integration test suite you wanted to use Postgres for metadata storage, but you (a) only have a single Postgres connection and (b) your base monad already has a `MonadError` instance. It would be nice to be able to reuse `PGMetadataStorageT` to select the Postgres-backed metadata implementation and just have that use the “acquire a Postgres connection and run a transaction” and “throw errors” capabilities of the underlying monad, whatever they happen to be.

If you require a `MonadError` constraint on the `MonadMetadataStorage` instance, then call sites can easily satisfy it by adding an extra `runExceptT` call around `runPGMetadataStorageT`. The `ExceptT` part doesn’t have to be embedded into the base monad for that to work.

The current approach in this PR seems very *direct*: the reasoning seems to be “we needed a monad transformer to hang aa `MonadMetadataStorage` instance on, and that instance needed error handling, so we might as well put the error handling in that monad transformer.” In isolation, that reasoning isn’t necessarily bad, but it’s totally at odds with the idea that `MetadataStorageT` is supposed to be a *general-purpose* transformer for hanging many different `MonadMetadataStorage` instances on. Even ignoring the `ExceptT` detail, the idiomatic `mtl`-style approach is to have one typeclass (which is the interface) and a transformer per implementation. This PR does something unidiomatic, which seems ill-advised to me unless you have a really good reason for it.

Reply:
@lexi-lambda We do have `MonadMetadataStorage (MetadataStorageT m)` as type class constraint for the functions in which the type class methods being called (See https://github.com/hasura/graphql-engine/blob/74c04362322a15fc569b3b9d826b894f813bb878/server/src-lib/Hasura/Eventing/ScheduledTrigger.hs). The idiomatic `mtl`-style is possible if we some how declare those functions as `MonadMetadataStorage m`.

Reply:
@lexi-lambda Addressed this conversation in a source code note added via https://github.com/hasura/graphql-engine/pull/6131/commits/b9b97eff670cf2708dc95743c921d2f369be79ca.